packages/tools/devtools/devtools-core/src/data-visualization/SharedTreeVisualizer.ts

```diff
@@ -0,0 +1,330 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+import type {
+	FieldMapObject,
+	JsonableTree,
+	SharedTreeContentSnapshot,
+	TreeFieldStoredSchema,
+	TreeNodeStoredSchema,
+} from "@fluidframework/tree/internal";
+import {
+	EmptyKey,
+	LeafNodeStoredSchema,
+	MapNodeStoredSchema,
+	ObjectNodeStoredSchema,
+} from "@fluidframework/tree/internal";
+import type { SharedTreeLeafNode, VisualSharedTreeNode } from "./VisualSharedTreeTypes.js";
+import { SharedTreeSchemaType, VisualSharedTreeNodeKind } from "./VisualSharedTreeTypes.js";
+import { type VisualChildNode, VisualNodeKind, type VisualValueNode } from "./VisualTree.js";
+
+/**
+ * Converts the output of {@link sharedTreeVisualizer} to {@link VisualChildNode} type containing `schema` and `children` fields.
+ */
+export function toVisualTree(tree: VisualSharedTreeNode): VisualChildNode {
+	if (tree.kind === VisualSharedTreeNodeKind.LeafNode) {
+		const result: VisualValueNode = {
+			value: tree.value,
+			nodeKind: VisualNodeKind.ValueNode,
+			tooltipContents: tree.schema.allowedTypes,
+		};
+		return result;
+	} else {
+		const children: Record<string, VisualChildNode> = {};
+
+		for (const [key, value] of Object.entries(tree.fields)) {
+			const child = toVisualTree(value);
+			children[key] = child;
+		}
+
+		return {
+			children,
+			nodeKind: VisualNodeKind.TreeNode,
+			tooltipContents: tree.schema.allowedTypes,
+		};
+	}
+}
+
+/**
+ * Returns the allowed fields & types for the object fields.
+ */
+function getObjectAllowedTypes(schema: ObjectNodeStoredSchema): string {
```

Comment:
We have a few different functions here that read allowedTypes data from different kinds of nodes, and also does the string building. We could simplify this some if we instead had separate functions for each node kind that creates an array of allowed types, and a separate utility function that converts that to the string format.